math/vecops/inc/ROOT/RVec.hxx

```diff
@@ -271,160 +259,1076 @@ hpt->Draw();
+   /// std::length_error or calls report_fatal_error.
+   static void report_at_maximum_capacity();
+
+public:
+   size_t size() const { return fSize; }
+   size_t capacity() const { return Capacity; }
+
+   RVEC_NODISCARD bool empty() const { return !fSize; }
+
+   /// Set the array size to \p N, which the current array must have enough
+   /// capacity for.
+   ///
+   /// This does not construct or destroy any elements in the vector.
+   ///
+   /// Clients can use this in conjunction with capacity() to write past the end
+   /// of the buffer when they know that more elements are available, and only
+   /// update the size later. This avoids the cost of value initializing elements
+   /// which will only be overwritten.
+   void set_size(size_t N)
+   {
+      assert(N <= capacity());
+      fSize = N;
+   }
+};
+
+template <class T>
+using SmallVectorSizeType = typename std::conditional<sizeof(T) < 4 && sizeof(void *) >= 8, uint64_t, uint32_t>::type;
+
+/// Figure out the offset of the first element.
+template <class T, typename = void>
+struct SmallVectorAlignmentAndSize {
+   alignas(SmallVectorBase<SmallVectorSizeType<T>>) char Base[sizeof(SmallVectorBase<SmallVectorSizeType<T>>)];
+   alignas(T) char FirstEl[sizeof(T)];
+};
+
+/// This is the part of SmallVectorTemplateBase which does not depend on whether
+/// the type T is a POD. The extra dummy template argument is used by ArrayRef
+/// to avoid unnecessarily requiring T to be complete.
+template <typename T, typename = void>
+class SmallVectorTemplateCommon : public SmallVectorBase<SmallVectorSizeType<T>> {
+   using Base = SmallVectorBase<SmallVectorSizeType<T>>;
+
+   /// Find the address of the first element.  For this pointer math to be valid
+   /// with small-size of 0 for T with lots of alignment, it's important that
+   /// SmallVectorStorage is properly-aligned even for small-size of 0.
+   void *getFirstEl() const
+   {
+      return const_cast<void *>(reinterpret_cast<const void *>(reinterpret_cast<const char *>(this) +
+                                                               offsetof(SmallVectorAlignmentAndSize<T>, FirstEl)));
+   }
+   // Space after 'FirstEl' is clobbered, do not add any instance vars after it.
+
+protected:
+   SmallVectorTemplateCommon(size_t Size) : Base(getFirstEl(), Size) {}
+
+   void grow_pod(size_t MinSize, size_t TSize) { Base::grow_pod(getFirstEl(), MinSize, TSize); }
+
+   /// Return true if this is a smallvector which has not had dynamic
+   /// memory allocated for it.
+   bool isSmall() const { return this->BeginX == getFirstEl(); }
+
+   /// Put this vector in a state of being small.
+   void resetToSmall()
+   {
+      this->BeginX = getFirstEl();
+      this->fSize = this->Capacity = 0; // FIXME: Setting Capacity to 0 is suspect.
+   }
+
+public:
+   using size_type = size_t;
+   using difference_type = ptrdiff_t;
+   using value_type = T;
+   using iterator = T *;
```

Comment:
The issue in the PyROOT iteration comes due to this. `RVec::begin()` does not return an iterator, it returns a `T *`!

@eguiraud There's a design decision behind this?